Add `EpollSystem` implementation for JVM

mox692 opened the pull request:
This PR provides a JVM implementation of `EpollSystem`, for the latest polling system on cats-effect.

* Port [the existing Native implementation](https://github.com/typelevel/cats-effect/blob/fdb2a3e468a67c51a513aec758fed11caa44a405/core/native/src/main/scala/cats/effect/unsafe/EpollSystem.scala) for the JVM.
* We will also add the necessary FFI interface (we'll use JNR).


Similar work has been done previously on work [fs2-io_uring project](https://github.com/armanbilge/fs2-io_uring/pull/78).

To avoid PR getting too large, I plan to address fs2 `Network` implementation in a separate PR.

mox692 commented:
@armanbilge 
Do we need to expose [FileDiscripterPoller](https://github.com/typelevel/cats-effect/blob/fdb2a3e468a67c51a513aec758fed11caa44a405/core/native/src/main/scala/cats/effect/FileDescriptorPoller.scala) to the JVM as well?
It seems that cats-effect currently publishes FileDiscripterPoller only in Native.

armanbilge commented:
In the long-term maybe we want to do that. In the short term we can just include that interface in this project, especially during the prototype stage. How does that sound?

mox692 commented:
Okay, I'll include that interface in this project.

mox692 commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,255 @@
+        cb: Either[Throwable, (PollHandle, IO[Unit])] => Unit
+    ): Unit = {
+      val events = (EPOLLET | (if (reads) EPOLLIN else 0) | (if (writes) EPOLLOUT else 0)).toInt
+      val data = (System.identityHashCode(handle).toLong << 32) | fd
+      val event = EpollEvent(events, data, globalRuntime)
+
+      val result =
+        if (epoll_ctl(epfd, EPOLL_CTL_ADD, fd, event) != 0)
+          Left(new IOException(strerror(errno())))
+        else {
+          handles.put(data, handle)
+          val remove = IO {
+            handles.remove(data)
```
I also think that by rewriting the `PollHandle` class implementation to be able to interop with native with jnr, we could set the `PollHandle` pointer directly in epoll_event.data, like native does.

However, if this implementation does not change the semantics, I think this way would be simpler and would not break the existing `PollHandle` implementation.

In this implementation, the `Poller` has a it's own hashmap, and the hashcode of the registered `Pollhandle` is kept in that hashmap.

armanbilge replied:
> we could set the `PollHandle` pointer directly in epoll_event.data, like native does.

I don't think it's possible to access the value of a pointer on the JVM. But maybe I didn't understand your proposal.

> In this implementation, the `Poller` has a it's own hashmap

I think a map or set will always be necessary, because we need to keep live references to the `PollHandle`s so they are not not GCed

mox692 commented on epoll/src/main/scala/fs2/io/epoll/unsafe/libc.scala:
```diff
@@ -0,0 +1,105 @@
+      .create(classOf[jnrLibcFFIInterface])
+      .load(Platform.getNativePlatform().getStandardCLibraryName())
+
+    /*
+     *  epoll_event layout
+     *
+     *  +---------------------+---------------------+---------------------+
+     *  |       events        |       padding       |       data          |
+     *  |       32bit         |       32bit         |       64bit         |
+     *  +---------------------+---------------------+---------------------+
+     *  Total size: 16 bytes
+     */
+    final class EpollEvent(runtime: Runtime) extends Struct(runtime) {
+      final private val _events: Unsigned32 = new Unsigned32()
+      final private val _data: Unsigned64 = new Unsigned64()
+
+      final val events: Int = _events.get().toInt
+      final val data: Long = _data.get()
+    }
```
EpollEvent was mapped like this on my linux machine. (but I don't know if all linux machines will have the same layout, so I need to check it later.)

armanbilge replied:
> all linux machines will have the same layout

Unfortunately not. The layout is different on x86_64 and ARM so we will need to account for that.

mox692 reviewed (commented):
@armanbilge @antoniojimeneznieto 
I left a few comments about implementation. There are still some rough spots or things to do, but could you check if I'm on the right track?

In fact, so far I was able to port almost directly from the native version `EpollSystem` in cats-effect. The only difference is the part where the epoll API is used, which is related to FFI.

----
By the way, I'm trying to write a test about EpollSystem, but the following simple IO program hangs🤔

```scala
abstract class EpollSuite extends CatsEffectSuite {
  override lazy val munitIORuntime =
    IORuntimeBuilder()
      .setPollingSystem(EpollSystem)
      .build()

  override def munitIOTimeout = 20.seconds
}

...

class EpollSystemSuite extends EpollSuite {
  test("sample suite") {
    IO(42).assertEquals(42)
  }
}
```

It appeared that many worker threads were blocking at the following location:

https://github.com/armanbilge/fs2-epoll/blob/b0418d75b37fea8ea413f921514c4fd253825167/epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala#L182

where they are called by `poll` method from worker threads with a timeout argument of `-1`, which cause `epoll_wait` block indefinitely.

Am I missing something?

armanbilge commented on epoll/src/main/scala/cats/effect/FileDescriptorPoller.scala:
```diff
@@ -0,0 +1,64 @@
+ * limitations under the License.
+ */
+
+package cats.effect
```
We should put this in the `fs2.epoll` package and keep it a private implementation detail for now, similar to `Uring` in fs2-io_uring.

antoniojimeneznieto commented:
Hi @mox692 😃 I've just checked it out and it is looking really really great!! :)

> where they are called by poll method from worker threads with a timeout argument of -1, which cause epoll_wait block indefinitely.

Hmmm sounds like there is a problem with the interrupt. In io_uring we used as a first idea a pipe, maybe we could use it here. Another option that we thought was to use an eventfd 🤔 What do you think @armanbilge ?

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,255 @@
+      readMutex: Mutex[IO],
+      writeMutex: Mutex[IO]
+  ) extends FileDescriptorPollHandle {
+    private[this] var readReadyCounter = 0
+    private[this] var readCallback: Either[Throwable, Int] => Unit = null
+
+    private[this] var writeReadyCounter = 0
+    private[this] var writeCallback: Either[Throwable, Int] => Unit = null
```
Some/all of these need to be `@volatile` in multi-threading. I need to think about it, for now we can mark them all `@volatile`.

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,255 @@
+
+  def needsPoll(poller: Poller): Boolean = poller.needsPoll()
+
+  def interrupt(targetThread: Thread, targetPoller: Poller): Unit = ()
```
Antonio is right, we definitely need interruption in multi-threading.

You can check how Antonio did it with a pipe, but I think in this case we should prefer to use an `eventfd`. An `eventfd` is kind of similar to FS2 `SignallingRef`: it is an atomic counter that you can wait for events on.

https://man7.org/linux/man-pages/man2/eventfd.2.html

The basic idea is:
1. We create an `eventfd` for each poller.
2. We should setup a read-ready event on the epoll for that `eventfd`. It will be read-ready whenever the value of the eventfd becomes greater than 0.
3. We can implement interruption for the poller by incrementing the value of the `eventfd` by writing to it. This will cause it to be greater than 0, and thus wake it up if it is polling.

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,255 @@
+
+  final class Poller private[EpollSystem] (epfd: Int) {
+
+    private[this] val handles: HashMap[Long, PollHandle] = HashMap()
```
We should use a Scala `LongMap` instead of a `HashMap`.

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,255 @@
+        cb: Either[Throwable, (PollHandle, IO[Unit])] => Unit
+    ): Unit = {
+      val events = (EPOLLET | (if (reads) EPOLLIN else 0) | (if (writes) EPOLLOUT else 0)).toInt
+      val data = (System.identityHashCode(handle).toLong << 32) | fd
```
Hmm, not sure about this. `identityHashCode` is not guaranteed to be unique: two different handles may have the same one.

I think it is sufficient to use the `fd` to identify the handle. The expectation is that the user should not register the same `fd` more than once.

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,255 @@
+        else {
+          handles.put(data, handle)
+          val remove = IO {
+            handles.remove(data)
+            if (epoll_ctl(epfd, EPOLL_CTL_DEL, fd, null) != 0)
+              throw new IOException(strerror(errno()))
```
I'm not sure if this is concurrency safe. The handle could be deregistered from a different thread.
- updating the map is definitely not concurrency safe. maybe for now we should use a `ConcurrentHashMap`
- I'm not sure if it is safe to call `epoll_ctl` from another thread. we should check about that.

armanbilge replied:
Yes, it looks like `epoll_ctl` is threadsafe, even when the that epoll instance is in an `epoll_wait`.

- https://man7.org/linux/man-pages/man2/epoll_wait.2.html#NOTES
- https://lkml.org/lkml/2006/2/28/367

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,255 @@
+  }
+
+  private object epoll {
+    final val EPOLL_CTL_ADD = 1
```
I wonder if some of these are in the JNR constants library.

armanbilge commented on build.sbt:
```diff
@@ -25,6 +25,7 @@ ThisBuild / githubWorkflowPublishPreamble +=
 val ceVersion = "3.6-e9aeb8c"
 val fs2Version = "3.9.2"
 val jnrFfiVersion = "2.2.15"
+val jnrConstatsVersion = "0.10.4"
```
```suggestion
val jnrConstantsVersion = "0.10.4"
```

armanbilge reviewed (commented):
First of all, very nice work!!

If you did a direct port of the Native version, then I'm not surprised that you are running into problems. Scala Native is not multi-threaded yet, so that implementation is missing several important things we need in this implementation.

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,255 @@
+              EpollEvent.apply(events, (i * EpollEvent.CStructSize).toLong, globalRuntime)
+            val handle = handles.get(epollEvent.data)
+
+            handle
+              .getOrElse(throw new IllegalStateException("Requested fd is not found"))
+              .notify(epollEvent.events.toInt)
```
I'm not sure if it makes sense to throw an exception here. The fd may not be found because that handle was canceled and there is a race condition.

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,290 @@
+    if (epfd == -1)
+      throw new IOException(strerror(errno()))
+
+    val evfd = libc.jnr.eventfd(0, EFD_SEMAPHORE | EFD_NONBLOCK)
```
Interesting, why did you add the `EFD_SEMAPHORE` flag? I not sure if its the behavior that we want.

mox692 replied:
Oh. I misunderstood the spec, you are correct.

If we set `EFD_SEMAPHORE`, the counter is decremented by 1 instead of being set to 0, so it is not suitable here.

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,290 @@
+    if (evfd == -1)
+      throw new IOException(strerror(errno()))
+
+    new Poller(epfd, evfd)
```
Sorry if I missed it, do we register the `evfd` with the `epfd` using `epoll_ctl` somewhere?

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,290 @@
+
+            i += 1
+          }
+        } else if (errno() != Errno.EINTR.intValue()) { // spurious wake-up by signal
```
I'm actually not sure if we need this on JVM. I guess it doesn't hurt.

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,290 @@
+        val triggeredEvents = epoll_wait(epfd, events, MaxEvents, timeout)
+
+        if (triggeredEvents >= 0) {
+          polled = true
+
+          var i = 0
+          while (i < triggeredEvents) {
+            val epollEvent =
+              EpollEvent.apply(events, (i * EpollEvent.CStructSize).toLong, globalRuntime)
+
+            if (triggeredEvents == 1 && epollEvent.data == evfd) {
+              // couldn't poll anything, we just received an interruption.
+              polled = false
+              return
+            }
+
+            val handle = handles.get(epollEvent.data)
+
+            // While polling, another worker thread may execute the PollHandle finalizer
+            // due to cancellation, and the target PollHandle may not exist in the `handles`.
+            // We execute `notify` method only if an entry exists in the HashMap.
+            if (handle ne null)
+              handle.notify(epollEvent.events.toInt)
```
```suggestion

          var i = 0
          while (i < triggeredEvents) {
            val epollEvent =
              EpollEvent.apply(events, (i * EpollEvent.CStructSize).toLong, globalRuntime)

            val handle = if (epollEvent.data != evfd) handles.get(epollEvent.data) else null

            // While polling, another worker thread may execute the PollHandle finalizer
            // due to cancellation, and the target PollHandle may not exist in the `handles`.
            // We execute `notify` method only if an entry exists in the HashMap.
            if (handle ne null) {
              handle.notify(epollEvent.events.toInt)
              polled = true
            }
```

mox692 commented:
@armanbilge @antoniojimeneznieto 

Thanks for the review, sorry for the delay in replying.
I have added the interruption implementation you pointed out, and added [some tests with a simple Pipe](https://github.com/armanbilge/fs2-epoll/blob/dd49878436d491b5954746babb6f43c5aed45558/epoll/src/test/scala/fs2/io/epoll/EpollSystemSuite.scala).

* Are there any other tests that should be added?
* Originally I was thinking of making `Network` a separate PR, but I think I will include the `Network` implementation in this PR, what do you think? (If there is a bug in the EpollSystem, I can be fix in this PR.)

Of course, other feedbacks are welcome.

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,300 @@
+
+  def close(): Unit = ()
+
+  def makeApi(register: (Poller => Unit) => Unit): Api =
+    new FileDescriptorPollerImpl(register)
```
```suggestion
  def makeApi(access: (Poller => Unit) => Unit): Api =
    new FileDescriptorPollerImpl(access)
```

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,300 @@
+    targetPoller.interrupt()
+
+  private final class FileDescriptorPollerImpl private[EpollSystem] (
+      register: (Poller => Unit) => Unit
```
```suggestion
      access: (Poller => Unit) => Unit
```

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,300 @@
+      Resource {
+        (Mutex[IO], Mutex[IO]).flatMapN { (readMutex, writeMutex) =>
+          IO.async_[(PollHandle, IO[Unit])] { cb =>
+            register { epoll =>
```
```suggestion
            access { epoll =>
```

armanbilge reviewed (commented):
Sorry, I haven't forgotten this PR! Thanks for all your work.

It looks like I started a review but then I got busy. I'll submit those comments and then I promise I'll try to continue my review soon.

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,300 @@
+  final class Poller private[EpollSystem] (epfd: Int, evfd: Int) {
+
+    // fixed array used only when reading from eventfd
+    private final lazy val readBuf = Memory.allocate(globalRuntime, 8)
```
We should avoid `lazy val` unless there is a strong motivating reason why it should be lazy. `lazy val`s add a memory synchronization penalty on every access.

```suggestion
    private[this] val readBuf = Memory.allocate(globalRuntime, 8)
```

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,300 @@
+
+    private[EpollSystem] def poll(timeout: Long): Boolean = {
+
+      val events = Memory.allocate(globalRuntime, EpollEvent.CStructSize * MaxEvents)
```
We should allocate this once instead of allocating every time we invoke `poll`.

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,300 @@
+            val epollEvent =
+              EpollEvent.apply(events, (i * EpollEvent.CStructSize).toLong, globalRuntime)
+
+            val handle = if (epollEvent.fd() != evfd) {
+              // We poll something.
+              handles.get(epollEvent.fd())
```
```suggestion
            val fd = epollEvent.fd()
            val handle = if (fd != evfd) {
              // We poll something.
              handles.get(fd)
```

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,300 @@
+
+          var i = 0
+          while (i < triggeredEvents) {
+            val epollEvent =
+              EpollEvent.apply(events, (i * EpollEvent.CStructSize).toLong, globalRuntime)
```
If we pre-allocate and cache these I wonder if it will give any performance improvement.

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala:
```diff
@@ -0,0 +1,300 @@
+              handles.get(epollEvent.fd())
+            } else {
+              // We received an interruption.
+              // Initialize the eventfd counter to 0.
```
```suggestion
              // Reset the eventfd counter to 0.
```

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/libc.scala:
```diff
@@ -0,0 +1,123 @@
+private[epoll] object libc {
+
+  object jnr {
+    lazy val globalRuntime = Runtime.getSystemRuntime
+
+    private lazy val libc = LibraryLoader
```
Same here, do these really need to be lazy?

armanbilge commented on epoll/src/main/scala/fs2/io/epoll/unsafe/libc.scala:
```diff
@@ -0,0 +1,123 @@
+        ev
+      }
+
+      def apply(ptr: Pointer, offset: Long, runtime: Runtime): EpollEvent = {
```
I'm confused what this method is doing. Is it copying the event from one native memory to another?
